File: Downloads/CSR-System-Final/csr-final/commands/event.py
import discord
from discord.ext import commands
from discord import app_commands
from typing import Optional
import asyncio
import os
from utils.db import get_events, add_event


class EventCog(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    event_group = app_commands.Group(
        name='event',
        description='Guild event management',
        default_permissions=discord.Permissions(manage_events=True)
    )

    @event_group.command(name='list', description='View upcoming guild events')
    async def event_list(self, interaction: discord.Interaction):
        events = get_events()
        if not events:
            return await interaction.response.send_message('📅 No events scheduled yet.', ephemeral=True)

        embed = discord.Embed(title='📅 Upcoming Guild Events', color=0x2d7dd2)
        embed.set_footer(text='Champions of the Shattered Realm')
        embed.timestamp = discord.utils.utcnow()

        for e in events[-10:]:
            embed.add_field(
                name=f"🎉 {e['name']}",
                value=f"📅 **{e['date']}**{chr(10) + e['content'] if e.get('content') else ''}\n🔔 Notified: {e['notify']}",
                inline=False
            )
        await interaction.response.send_message(embed=embed)

    @event_group.command(name='create', description='Create an event and send DM reminders')
    @app_commands.describe(
        name='Event name',
        date='Event date & time (e.g. "July 20 at 8pm UTC")',
        content='Optional event description',
        notify='Who to notify',
        role='Specific role to notify (if "role" chosen)',
        ping='Also ping in announcements channel? (default: True)'
    )
    @app_commands.choices(notify=[
        app_commands.Choice(name='@everyone', value='everyone'),
        app_commands.Choice(name='Specific Role', value='role'),
    ])
    async def event_create(
        self,
        interaction: discord.Interaction,
        name: str,
        date: str,
        notify: str,
        content: Optional[str] = None,
        role: Optional[discord.Role] = None,
        ping: Optional[bool] = True,
    ):
        if notify == 'role' and not role:
            return await interaction.response.send_message(
                '❌ You chose "Specific Role" but didn\'t select a role. Run the command again.', ephemeral=True
            )

        await interaction.response.defer(ephemeral=True)

        # Save event
        event_data = {
            'name': name, 'date': date, 'content': content or '',
            'notify': role.name if notify == 'role' else '@everyone',
            'created_by': str(interaction.user),
        }
        add_event(event_data)

        guild = interaction.guild
        guild_name = os.getenv('GUILD_NAME', guild.name)

        dm_embed = discord.Embed(
            title=f'📅 Upcoming Event: {name}',
            description=(
                f'Hey! There\'s a new event in **{guild_name}** — don\'t miss it!\n\n'
                f'**📆 When:** {date}\n'
                + (f'**📝 Details:** {content}\n' if content else '')
                + '\nSee you there! 🎉'
            ),
            color=0x2d7dd2
        )
        dm_embed.set_footer(text='Champions of the Shattered Realm · Event Reminder')
        dm_embed.timestamp = discord.utils.utcnow()

        # Determine targets
        await guild.chunk()
        targets = [m for m in guild.members if not m.bot]
        if notify == 'role' and role:
            targets = [m for m in targets if role in m.roles]

        await interaction.followup.send(
            f'📤 Sending DMs to **{len(targets)}** member(s)... this may take a moment.',
            ephemeral=True
        )

        sent = failed = 0
        for member in targets:
            try:
                await member.send(embed=dm_embed)
                sent += 1
                await asyncio.sleep(0.3)  # rate-limit friendly
            except Exception:
                failed += 1

        # Optional channel ping
        if ping and os.getenv('WELCOME_CHANNEL_ID'):
            ch = guild.get_channel(int(os.getenv('WELCOME_CHANNEL_ID')))
            if ch:
                ping_str = '@everyone' if notify == 'everyone' else role.mention
                try:
                    await ch.send(
                        content=f'{ping_str} 📅 **New Event: {name}** — {date}' + (f' · {content}' if content else ''),
                        embed=dm_embed,
                        allowed_mentions=discord.AllowedMentions(everyone=True, roles=True)
                    )
                except Exception:
                    pass

        await interaction.followup.send(
            f'✅ Event created! DMs sent: **{sent}** · Failed (DMs closed): **{failed}**',
            ephemeral=True
        )


async def setup(bot):
    cog = EventCog(bot)
    bot.tree.add_command(cog.event_group)
    await bot.add_cog(cog)

File: Downloads/champions-guild-bot/commands/event.js
const { SlashCommandBuilder, EmbedBuilder, PermissionFlagsBits } = require('discord.js');
const { db } = require('../utils/db');

module.exports = {
  data: new SlashCommandBuilder()
    .setName('event')
    .setDescription('Guild event management')
    .setDefaultMemberPermissions(PermissionFlagsBits.ManageEvents)

    .addSubcommand(sub => sub.setName('create')
      .setDescription('Create an event and send DM reminders')
      .addStringOption(o => o.setName('name').setDescription('Event name').setRequired(true))
      .addStringOption(o => o.setName('date').setDescription('Event date & time (e.g. "July 20 at 8pm UTC")').setRequired(true))
      .addStringOption(o => o.setName('content').setDescription('Optional event description'))
      .addStringOption(o => o.setName('notify')
        .setDescription('Who to notify?').setRequired(true)
        .addChoices(
          { name: '@everyone', value: 'everyone' },
          { name: 'Specific Role', value: 'role' },
        )
      )
      .addRoleOption(o => o.setName('role').setDescription('Role to notify (if "Specific Role" chosen)'))
      .addBooleanOption(o => o.setName('ping').setDescription('Also ping in announcements channel? (default: true)'))
    )
    .addSubcommand(sub => sub.setName('list')
      .setDescription('View all upcoming events')
    ),

  async execute(interaction) {
    const sub = interaction.options.getSubcommand();

    if (sub === 'list') {
      const events = db.getEvents();
      if (!events.length) {
        return interaction.reply({ content: '📅 No events scheduled yet.', ephemeral: true });
      }
      const embed = new EmbedBuilder()
        .setColor(0x2d7dd2)
        .setTitle('📅 Upcoming Guild Events')
        .setTimestamp()
        .setFooter({ text: 'Champions of the Shattered Realm' });

      events.slice(-10).forEach(e => {
        embed.addFields({
          name: `🎉 ${e.name}`,
          value: `📅 **${e.date}**${e.content ? `\n${e.content}` : ''}\n🔔 Notified: ${e.notify}`,
        });
      });
      return interaction.reply({ embeds: [embed] });
    }

    if (sub === 'create') {
      await interaction.deferReply({ ephemeral: true });

      const name    = interaction.options.getString('name');
      const date    = interaction.options.getString('date');
      const content = interaction.options.getString('content') || '';
      const notify  = interaction.options.getString('notify');
      const role    = interaction.options.getRole('role');
      const ping    = interaction.options.getBoolean('ping') ?? true;

      if (notify === 'role' && !role) {
        return interaction.editReply('❌ You chose "Specific Role" but didn\'t select a role. Run the command again.');
      }

      // Save event
      const eventData = { name, date, content, notify: notify === 'role' ? role.name : '@everyone', createdBy: interaction.user.tag };
      db.addEvent(eventData);

      // Build DM embed
      const dmEmbed = new EmbedBuilder()
        .setColor(0x2d7dd2)
        .setTitle(`📅 Upcoming Event: ${name}`)
        .setDescription(
          `Hey! There's a new event in **${process.env.GUILD_NAME || 'the guild'}** — don't miss it!\n\n` +
          `**📆 When:** ${date}\n` +
          (content ? `**📝 Details:** ${content}\n` : '') +
          `\nSee you there! 🎉`
        )
        .setFooter({ text: 'Champions of the Shattered Realm · Event Reminder' })
        .setTimestamp();

      // Determine who to DM
      const guild = interaction.guild;
      await guild.members.fetch();

      let targets = guild.members.cache.filter(m => !m.user.bot);
      if (notify === 'role' && role) {
        targets = targets.filter(m => m.roles.cache.has(role.id));
      }

      let sent = 0, failed = 0;
      await interaction.editReply(`📤 Sending DMs to **${targets.size}** member(s)... this may take a moment.`);

      for (const member of targets.values()) {
        try {
          await member.send({ embeds: [dmEmbed] });
          sent++;
          // Rate limit friendly — small delay
          await new Promise(r => setTimeout(r, 300));
        } catch { failed++; }
      }

      // Optionally ping in announcements channel
      if (ping && process.env.WELCOME_CHANNEL_ID) {
        const ch = guild.channels.cache.get(process.env.WELCOME_CHANNEL_ID);
        if (ch) {
          const pingText = notify === 'everyone' ? '@everyone' : `<@&${role.id}>`;
          await ch.send({ content: `${pingText} 📅 **New Event: ${name}** — ${date}${content ? ` · ${content}` : ''}`, embeds: [dmEmbed] }).catch(() => {});
        }
      }

      return interaction.editReply(`✅ Event created! DMs sent: **${sent}** · Failed (DMs closed): **${failed}**`);
    }
  },
};
